File: src/brm/filesystem.cpp
#include "filesystem.h"

#include <sys/stat.h>    // stat

using namespace std;

//Stats a path, returns the kind of object.
stat_result lookup_path(const string& path)
{
	struct stat s;

	if (stat(path.c_str(), &s) == -1) {
		if (errno != ENOENT) {
			perror("stat");
			exit(1);
		}
		return Nothing;
	}
	if (S_ISDIR(s.st_mode)) {
		return Directory;
	} else if (S_ISREG(s.st_mode)) {
		return File;
	}
	return Other;
}

File: src/brm/Filesystem.cs
/*
 * Filesystem.cs
 *
 *      This program is free software; you can redistribute it and/or
 *      modify it under the terms of the GNU General Public License
 *      version 2 as published by the Free Software Foundation.
 *
 * Copyright (c) 2014 Leonardo Taglialegne <cmt.miniBill@gmail.com>
 *
 */

using Mono.Unix;

namespace Brm
{
	public static class Filesystem
	{
		/// <summary>
		/// Stats a path, returns the kind of object.
		/// </summary>
		/// <returns>The kind of object.</returns>
		/// <param name="path">A path.</param>
		public static StatResult LookupPath (string path)
		{
			UnixFileSystemInfo entry = UnixFileSystemInfo.GetFileSystemEntry (path);
			if (!entry.Exists)
				return StatResult.Nothing;
			switch (entry.FileType) {
			case FileTypes.RegularFile:
				return StatResult.File;
			case FileTypes.Directory:
				return StatResult.Directory;
			default:
				return StatResult.Other;
			}
		}
	}
}